hypervisor/vm.go

```diff
@@ -158,171 +132,226 @@ func (vm *Vm) ReleaseVm() (int, error) {
 	if err != nil {
-		return errorResponse(err.Error())
+		vm.ctx.Log(ERROR, "fail to get response channel: %v", err)
+		return nil
 	}
-	defer vm.ReleaseResponseChan(Status)
 
-	go vm.handlePodEvent(mypod)
+	go func() {
+		defer vm.ReleaseResponseChan(Status)
+		for len(waiting) > 0 {
+			select {
+			case response, ok := <-Status:
+				if !ok {
+					vm.ctx.Log(WARNING, "status chan broken during waiting containers: %#v", waiting)
+					close(result)
+					return
+				}
+				if response.Code == waitEvent {
+					ps, _ := response.Data.(*types.ProcessFinished)
+					if _, ok := waiting[ps.Id]; ok {
+						result <- &api.ProcessExit{
+							Id:   ps.Id,
+							Code: int(ps.Code),
+							FinishedAt: time.Now().UTC(),
+						}
+					}
+				}
+			case <-timeoutChan:
+				vm.ctx.Log(WARNING, "timeout while waiting result of containers: %#v", waiting)
+				close(result)
+				return
+			}
+		}
+		close(result)
+	}()
+
+	return result
+}
 
-	runPodEvent := &RunPodCommand{
-		Spec:       userPod,
-		Containers: cList,
-		Volumes:    vList,
-		Wg:         mypod.Wg,
+//func (vm *Vm) handlePodEvent(mypod *PodStatus) {
+//	glog.V(1).Infof("hyperHandlePodEvent pod %s, vm %s", mypod.Id, vm.Id)
+//
+//	Status, err := vm.GetResponseChan()
+//	if err != nil {
+//		return
+//	}
+//	defer vm.ReleaseResponseChan(Status)
+//
+//	exit := false
+//	mypod.Wg.Add(1)
+//	for {
+//		Response, ok := <-Status
+//		if !ok {
+//			break
+//		}
+//
+//		switch Response.Code {
+//		case types.E_CONTAINER_FINISHED:
+//			ps, ok := Response.Data.(*types.ProcessFinished)
+//			if ok {
+//				mypod.SetOneContainerStatus(ps.Id, ps.Code)
+//				close(ps.Ack)
+//			}
+//		case types.E_EXEC_FINISHED:
+//			ps, ok := Response.Data.(*types.ProcessFinished)
+//			if ok {
+//				mypod.SetExecStatus(ps.Id, ps.Code)
+//				close(ps.Ack)
+//			}
+//		case types.E_POD_FINISHED: // successfully exit
+//			mypod.SetPodContainerStatus(Response.Data.([]uint32))
+//			//vm.Status = types.S_VM_IDLE
+//		case types.E_VM_SHUTDOWN: // vm exited, sucessful or not
+//			if mypod.Status == types.S_POD_RUNNING { // not received finished pod before
+//				mypod.Status = types.S_POD_FAILED
+//				mypod.FinishedAt = time.Now().Format("2006-01-02T15:04:05Z")
+//				mypod.SetContainerStatus(types.S_POD_FAILED)
+//			}
+//			mypod.Vm = ""
+//			exit = true
+//		}
+//
+//		if mypod.Handler != nil {
+//			mypod.Handler.Handle(Response, mypod.Handler.Data, mypod, vm)
+//		}
+//
+//		if exit {
+//			vm.clients = nil
+//			break
+//		}
+//	}
+//	mypod.Wg.Done()
+//}
+
+func (vm *Vm) InitSandbox(config *api.SandboxConfig) api.Result {
+	if vm.ctx == nil {
+		return NewNotReadyError(vm.Id)
+	}
+
+	vm.ctx.SetNetworkEnvironment(config)
+	Status, err := vm.GetResponseChan()
+	if err != nil {
+		vm.ctx.Log(ERROR, "failed to get status chan to monitor startpod: %v", err)
+		return api.NewResultBase(vm.Id, false, err.Error())
 	}
+	defer vm.ReleaseResponseChan(Status)
 
-	vm.Hub <- runPodEvent
+	vm.ctx.startPod()
```

Comment:
Call ctx.startPod directly here, If there is no need to handle runpod Event in state machine?